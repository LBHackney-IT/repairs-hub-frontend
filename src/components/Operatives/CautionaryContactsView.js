import BackButton from '../Layout/BackButton'
import { Table, THead, TBody, TR, TH, TD } from '../Layout/Table'
import { CAUTIONARY_CONTACT } from '../../utils/cautContactAlertsDescription'
const CautionaryContactView = () => (
  <>
    <BackButton />
    <h3 className="lbh-heading-h3">Cautionary contact</h3>
    <Table>
      <THead>
        <TR className="lbh-body">
          <TH scope="col">Code</TH>
          <TH scope="col">Description</TH>
        </TR>
      </THead>
      <TBody>
        {CAUTIONARY_CONTACT.map((entry, index) => (
          <TR key={index} index={index}>
            <TD>{entry.code}</TD>
            <TD>{entry.description}</TD>
          </TR>
        ))}
      </TBody>
    </Table>
  </>
)

export default CautionaryContactView
