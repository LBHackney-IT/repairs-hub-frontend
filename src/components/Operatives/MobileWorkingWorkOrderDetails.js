import PropTypes from 'prop-types'
import { WorkOrder } from '@/models/workOrder'
import { GridColumn, GridRow } from '../Layout/Grid'
import { useRouter } from 'next/router'
import { getCautionaryAlertsType } from '@/utils/cautionaryAlerts'
import TruncateText from '../Layout/TruncateText'

const MobileWorkingWorkOrderDetails = ({
  property,
  workOrder,
  personAlerts,
  locationAlerts,
}) => {
  const router = useRouter()
  const cautionaryAlertsType = getCautionaryAlertsType(
    locationAlerts,
    personAlerts
  )

  const cautContactURL = () => {
    router.push({
      pathname: `/work-orders/cautionary-alerts`,
      query: {
        cautContactCodes: cautionaryAlertsType,
      },
    })
  }

  return (
    <>
      <div className="operative-work-order">
        <h1 className="lbh-heading-h1">
          WO&nbsp;{workOrder.reference.toString().padStart(8, '0')}
        </h1>
        <div className="priority-text govuk-!-margin-top-0">
          {workOrder.isHigherPriority() ? (
            <div className="text-dark-red lbh-heading-h3">
              {' '}
              {workOrder.priority.toLowerCase().split(' ').slice(-1)}
            </div>
          ) : (
            <div className="lbh-heading-h3">
              {' '}
              {workOrder.priority.toLowerCase().split(' ').slice(-1)}
            </div>
          )}
        </div>
        <div className="lbh-heading-h4">Description</div>
        <TruncateText
          text={workOrder.description}
          numberOfLines="3"
          pTagClassName={'govuk-body govuk-!-margin-bottom-0'}
          linkClassName={'govuk-body'}
        ></TruncateText>

        {workOrder.plannerComments && (
          <>
            <div className="lbh-heading-h5">Planner Comment</div>
            <p className="govuk-body">{workOrder.plannerComments}</p>
          </>
        )}
        <div className="work-order-information">
          {cautionaryAlertsType && (
            <GridRow className="align-grid-row">
              <GridColumn width="one-half">
                <a
                  className="lbh-heading-h5 lbh-link"
                  href="#"
                  id="caut-alerts"
                  onClick={cautContactURL}
                >
                  Caut. alerts
                </a>
              </GridColumn>
              <GridColumn width="one-half" className="align-grid-column">
                <div className="govuk-warning-text lbh-warning-text">
                  <span
                    className="govuk-warning-text__icon person-alert--icon"
                    aria-hidden="true"
                  >
                    !
                  </span>
                  <strong className="govuk-warning-text__text person-alert--text">
                    {cautionaryAlertsType}
                  </strong>
                </div>
              </GridColumn>
            </GridRow>
          )}

          <GridRow className="align-grid-row">
            <GridColumn width="one-half">
              <div className="lbh-body property-name">Address</div>
            </GridColumn>
            <GridColumn width="one-half" className="align-grid-column">
              <div className="lbh-body">
                {property.address.shortAddress} {property.address.postalCode}
              </div>
            </GridColumn>
          </GridRow>

          {workOrder.callerName && (
            <GridRow className="align-grid-row">
              <GridColumn width="one-half">
                <div className="lbh-body property-name">Contact</div>
              </GridColumn>
              <GridColumn width="one-half" className="align-grid-column">
                <div className="lbh-body">{workOrder.callerName}</div>
              </GridColumn>
            </GridRow>
          )}

          {workOrder.callerNumber && (
            <GridRow className="align-grid-row">
              <GridColumn width="one-half">
                <div className="lbh-body property-name">Telephone no</div>
              </GridColumn>
              <GridColumn width="one-half" className="align-grid-column">
                <div className="lbh-body">{workOrder.callerNumber}</div>
              </GridColumn>
            </GridRow>
          )}

          {workOrder.appointment && workOrder.appointment.note && (
            <GridRow className="align-grid-row">
              <GridColumn width="one-half">
                <div className="lbh-body property-name">Comment</div>
              </GridColumn>
              <GridColumn width="one-half" className="align-grid-column">
                <div className="lbh-body">{workOrder.appointment.note}</div>
              </GridColumn>
            </GridRow>
          )}
        </div>
      </div>
    </>
  )
}

MobileWorkingWorkOrderDetails.propTypes = {
  property: PropTypes.object.isRequired,
  workOrder: PropTypes.instanceOf(WorkOrder).isRequired,
  personAlerts: PropTypes.array.isRequired,
  locationAlerts: PropTypes.array.isRequired,
}

export default MobileWorkingWorkOrderDetails
