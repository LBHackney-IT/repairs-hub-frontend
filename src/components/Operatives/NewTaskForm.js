import { useForm } from 'react-hook-form'
import { useState, useEffect } from 'react'
import { useRouter } from 'next/router'
import { frontEndApiRequest } from '@/utils/frontEndApiClient/requests'
import { PrimarySubmitButton } from '../Form'
import BackButton from '../Layout/BackButton'
import RateScheduleItem from '../WorkElement/RateScheduleItem'
import Spinner from '../Spinner'
import ErrorMessage from '../Errors/ErrorMessage'
import { buildVariationFormData } from '@/utils/hact/jobStatusUpdate/variation'
import { getWorkOrder } from '../../utils/requests/workOrders'
import { APIResponseError } from '../../types/requests/types'
import { formatRequestErrorMessage } from '../../utils/errorHandling/formatErrorMessage'

const NewTaskForm = ({ workOrderReference }) => {
  const [loading, setLoading] = useState(false)
  const [error, setError] = useState()
  const [currentUser, setCurrentUser] = useState({})
  const [sorCodes, setSorCodes] = useState([])
  const [existingTasks, setExistingTasks] = useState({})
  const { register, handleSubmit, errors } = useForm()
  const router = useRouter()

  const getNewTaskForm = async (reference) => {
    setError(null)

    try {
      const workOrderResponse = await getWorkOrder(reference)

      if (!workOrderResponse.success) {
        throw workOrderResponse.error
      }

      const workOrder = workOrderResponse.response

      const sorCodes = await frontEndApiRequest({
        path: '/api/schedule-of-rates/codes',
        method: 'get',
        params: {
          tradeCode: workOrder.tradeCode,
          propertyReference: workOrder.propertyReference,
          contractorReference: workOrder.contractorReference,
          showAdditionalTrades: true,
        },
      })

      setSorCodes(sorCodes)

      const currentUser = await frontEndApiRequest({
        method: 'get',
        path: '/api/hub-user',
      })

      setCurrentUser(currentUser)

      const tasksAndSors = await frontEndApiRequest({
        method: 'get',
        path: `/api/workOrders/${workOrderReference}/tasks`,
      })

      setExistingTasks(tasksAndSors)
    } catch (e) {
      setSorCodes([])

      if (e instanceof APIResponseError) {
        setError(e.message)
      } else {
        setError(formatRequestErrorMessage(e))
      }
    }

    setLoading(false)
  }

  const onFormSubmit = async (e) => {
    setLoading(true)

    const newTask = [
      {
        ...e.rateScheduleItems.operative,
        description: e.rateScheduleItems.operative.code.split(' - ')[1],
      },
    ]

    const workOrderUpdateFormData = buildVariationFormData(
      existingTasks,
      newTask,
      workOrderReference
    )

    try {
      await frontEndApiRequest({
        method: 'post',
        path: `/api/jobStatusUpdate`,
        requestData: workOrderUpdateFormData,
      })

      router.push(
        `/operatives/${currentUser?.operativePayrollNumber}/work-orders/${workOrderReference}`
      )
    } catch (e) {
      console.error(e)

      setError(formatRequestErrorMessage(e))
    }

    setLoading(false)
  }

  useEffect(() => {
    setLoading(true)
    setError(null)

    getNewTaskForm(workOrderReference)
  }, [])

  return (
    <>
      {loading ? (
        <Spinner />
      ) : (
        <>
          <BackButton />
          {sorCodes && (
            <>
              <h1 className="lbh-heading-h2 govuk-!-margin-bottom-4">
                New SOR
              </h1>
              <form
                role="form"
                id="repair-request-form"
                onSubmit={handleSubmit(onFormSubmit)}
              >
                <RateScheduleItem
                  sorCodes={sorCodes}
                  register={register}
                  showRemoveRateScheduleItem={false}
                  removeRateScheduleIte={false}
                  errors={errors}
                  index={'operative'}
                />
                <div className="button-pair">
                  <PrimarySubmitButton
                    id="submit-work-order-task-add"
                    label="Confirm"
                  />
                </div>
              </form>
            </>
          )}
          {error && <ErrorMessage label={error} />}
        </>
      )}
    </>
  )
}

export default NewTaskForm
