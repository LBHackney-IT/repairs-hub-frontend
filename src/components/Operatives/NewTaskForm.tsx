import WorkOrderUpdateView from '../WorkOrder/Update'
import { mobileViewUpdateWorkOrderLinks } from '../../utils/successPageLinks'

interface Props {
  workOrderReference: string
  operativePayrollNumber: string
}

const NewTaskForm = (props: Props) => {
  const { workOrderReference, operativePayrollNumber } = props

  return (
    <WorkOrderUpdateView
      reference={workOrderReference}
      // isMobileView={true}
      mobileViewLinks={mobileViewUpdateWorkOrderLinks(
        operativePayrollNumber,
        workOrderReference
      )}
    />
  )
}

export default NewTaskForm
