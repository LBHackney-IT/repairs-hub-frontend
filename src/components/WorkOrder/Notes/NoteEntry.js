import PropTypes from 'prop-types'
import { formatDateTime } from '@/utils/time'

const IMAGE_DETECTION_STRING = 'Image from user: '

const NoteInfo = ({ time, user, email }) => {
  return (
    <>
      {time ? formatDateTime(time) : 'â€”'}{' '}
      {user && email ? `by ${user} (${email})` : ''}
    </>
  )
}

const DetectImageNote = (note) => {
  return note.includes(IMAGE_DETECTION_STRING)
}

const ParseUriFromImageNote = (note) => {
  return note.replace(IMAGE_DETECTION_STRING, '')
}

const NoteEntry = ({ note, time, user, userEmail }) => {
  return (
    <>
      <div className="note-info lbh-body-s">
        <NoteInfo time={time} user={user} email={userEmail} />
      </div>
      {DetectImageNote(note) ? (
        <a href={ParseUriFromImageNote(note)}>{'Image from user'}</a>
      ) : (
        note.split('\n').map((el, i) => (
          <span key={i}>
            {el}
            <br />
          </span>
        ))
      )}
    </>
  )
}

NoteEntry.propTypes = {
  note: PropTypes.string.isRequired,
  time: PropTypes.instanceOf(Date).isRequired,
  user: PropTypes.string,
}

export default NoteEntry
