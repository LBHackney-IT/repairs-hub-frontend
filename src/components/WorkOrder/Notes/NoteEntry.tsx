import { formatDateTime } from '@/utils/time'
import { Note } from './types'
import NormalNoteContent from './NoteContent/NormalNoteContent'
import NoteWithImageContent from './NoteContent/NoteWithImageContent'
import CompletedNoteContent from './NoteContent/CompletedNoteContent'
import { WorkOrder } from '@/root/src/models/workOrder'
import { TabName } from '../../Tabs/types'

const NoteInfo = ({ note }: { note: Note }) => {
  const { time, user, userEmail } = note

  return (
    <>
      {time ? formatDateTime(time) : 'â€”'}{' '}
      {user && userEmail ? `by ${user} (${userEmail})` : ''}
    </>
  )
}

const IMAGE_DETECTION_STRING = 'Image from user: '

const DetectImageNote = (userNote: string) => {
  if (userNote === null) return false
  return userNote.includes(IMAGE_DETECTION_STRING)
}

interface Props {
  note: Note
  workOrder: WorkOrder
  setActiveTab: (tab: string) => void
}

const NoteEntry = ({ note, workOrder, setActiveTab }: Props) => {
  return (
    <>
      <div className="note-info lbh-body-s">
        <NoteInfo note={note} />
      </div>

      <NoteContent
        note={note}
        workOrder={workOrder}
        setActiveTab={setActiveTab}
      />
    </>
  )
}

const NoteContent = ({
  note,
  workOrder,
  setActiveTab,
}: {
  note: Note
  workOrder: WorkOrder
  setActiveTab: (tabName: TabName) => void
}) => {
  if (
    note.noteGeneratedOnFrontend &&
    note.typeCode.toString() === '0' &&
    note.otherType === 'completed'
  ) {
    return (
      <CompletedNoteContent
        note={note}
        workOrder={workOrder}
        setActiveTab={setActiveTab}
      />
    )
  }

  if (DetectImageNote(note.note)) {
    return <NoteWithImageContent note={note} />
  }

  return <NormalNoteContent note={note} />
}

export default NoteEntry
