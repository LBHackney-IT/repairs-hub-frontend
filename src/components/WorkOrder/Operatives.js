const Operatives = ({ operatives }) => {
  return (
    <p className="lbh-body-xs">
      {`${operatives.length > 1 ? 'Operatives' : 'Operative'}: ${operatives
        .map((operative) => operative.name)
        .join(', ')}`}
    </p>
  )
}

export default Operatives
