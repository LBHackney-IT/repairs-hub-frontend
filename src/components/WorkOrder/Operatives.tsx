import { Operative } from '../../models/operativeModel'

interface Props {
  operatives: Operative[]
}

const Operatives = ({ operatives }: Props) => {
  return (
    <p className="lbh-body-xs">
      {`${operatives.length > 1 ? 'Operatives' : 'Operative'}: ${operatives
        .map((operative) => operative.name)
        .join(', ')}`}
    </p>
  )
}

export default Operatives
