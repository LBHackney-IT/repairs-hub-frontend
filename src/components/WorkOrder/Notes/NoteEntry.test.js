import { render } from '@testing-library/react'
import NoteEntry from './NoteEntry'
import Operatives from '../Operatives'

describe('NoteEntry component', () => {
  it('should render note with image (repairs online)', () => {
    const props = {
      note: {
        note: 'Image from user: http://example-url/',
        time: new Date('2021-01-23T16:28:57.17811'),
        user: 'Test user',
        userEmail: 'sample@sample.com',
        noteGeneratedOnFrontend: false,
        typeCode: '3',
        otherType: 'some type',
      },
    }

    const { asFragment } = render(
      <NoteEntry note={props.note} workOrder={{}} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render a normal note', () => {
    const props = {
      note: {
        note: 'some pre-generated note',
        time: new Date('2021-01-23T16:28:57.17811'),
        user: 'Test user',
        userEmail: 'sample@sample.com',
        noteGeneratedOnFrontend: false,
        typeCode: '3',
        otherType: 'some type',
      },
    }

    const { asFragment } = render(
      <NoteEntry note={props.note} workOrder={{}} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render a completed work order note', () => {
    const props = {
      note: {
        note: 'comments left by user',
        time: new Date('2021-01-23T16:28:57.17811'),
        user: 'Test user',
        userEmail: 'sample@sample.com',
        noteGeneratedOnFrontend: true,
        typeCode: '0',
        otherType: 'completed',
      },
      workOrder: {
        closedDated: '2021-01-23T16:28:57.17811',
        operatives: [
          {
            name: 'Steve',
            jobPercentage: 15,
          },
        ],
        paymentType: 'Bonus',
      },
    }

    const { asFragment } = render(
      <NoteEntry note={props.note} workOrder={props.workOrder} />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
