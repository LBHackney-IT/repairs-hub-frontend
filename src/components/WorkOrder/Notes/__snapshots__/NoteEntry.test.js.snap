// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteEntry component should render properle 1`] = `
<DocumentFragment>
  <div
    class="note-info lbh-body-s"
  >
    20 Dec 2022, 15:32 by Test user (sample@sample.com)
  </div>
  <a
    href="http://example-url/"
  >
    Image from user
  </a>
</DocumentFragment>
`;
